csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,370 @@ class AllocationReuseModifier : private kir::ExprMutator {
+}
+
+Val* allocSizeBytes(kir::Allocate* alloc) {
+  const auto buffer_dtype = alloc->buffer()->dtype();
+  const auto dtype_size = dataTypeSize(buffer_dtype);
+  auto size = dtype_size == 1
+      ? alloc->size()
+      : SimplifyingIrBuilder::mulExpr(
+            alloc->size(), IrBuilder::create<Val>(dtype_size, DataType::Index));
+  return size;
+}
+
+//! Allocate differently-sized buffers using a single pass where we push
+//! allocations on a stack then pop them after their last read. This only does
+//! outer sharing: inner sharing is only valid for aliasing.
+//!
+//! Consider the following case, where time proceeds from left to right and
+//! memory is laid out in the upward direction by NoReuseSharedMemAllocator:
+//!
+//!                 +-----+
+//!                 |  C  |
+//!       +-----+   +-----+
+//!       |  B  |
+//!   +---+-----+-----+
+//!   |      A        |
+//!   +---------------+
+//!
+//! In this case the A allocation overlaps both B and C so it cannot be re-used
+//! but we can re-use B's memory. If B and C are compatible, they can be
+//! aliased; however, in this class we assume that all eligible aliases are
+//! already set. We can still reuse memory like below (time points are labelled
+//! along the horizontal axis):
+//!
+//!       +-----+   +-----+
+//!       |  B  |   |  C  |
+//!   +---+-----+-----+---+
+//!   |      A        |
+//!   +---------------+
+//!   a   b     c   d e   f
+//!
+//! Note that we might have incomplete information about allocation sizes prior
+//! to runtime, so our approach only places new allocations on top of the
+//! highest active previous allocation. We implement this with a stack of
+//! allocations to which we push and pop at each time point of interest:
+//!
+//!   a: Allocate A at address 0. Push A
+//!   b: Set T=stack.back()=A. Allocate B at address(T)+size(T). Push B
+//!   c: Pop B
+//!   d: Set T=stack.back()=A. Allocate C at address(T)+size(T). Push C
+//!   e: Don't pop A since it is not at back of stack.
+//!   f: Pop C and A (all inactive allocations at top of stack)
+//!
+//! [Reordering Pushes]
+//! Consider the following case:
+//!
+//!           +-----+
+//!           |  C  |
+//!       +---+-+---+
+//!       |  B  |
+//!   +---+-+---+
+//!   |  A  |
+//!   +-----+
+//!   a   b c d e   f
+//!
+//! A simple linear stack approach would fail to re-use any memory in the case
+//! illustrated above, even though A can be overwritten for C:
+//!
+//!   a: Push A
+//!   b: Push B
+//!   c: Cannot pop A since it is covered by B
+//!   d: Must allocate C on top of B, which is on top of A. i.e. NO RE-USE
+//!   e: Cannot pop B since it is covered by C
+//!   f: Finally pop C, B, and A
+//!
+//! This kind of case is actually quite common. For example, in a tiled matmul
+//! (ignoring double buffering) we have input tiles A and B, and an intermediate
+//! output tile C, which all overlap one another, then we might unswizzle the
```

Comment:
What is intermediate output tile C?

Reply:
I think for matmul, there are only three buffers, two for inputs, and one for result. The result buffer is the buffer for unswizzle, and there is no overlap for this buffer with the inputs buffer?

Reply:
Ah right. In that case if there's no overlap then reordering is not required for reusing both input buffers. I'll edit the comment.